internal/consensus/state.go

```diff
@@ -2053,6 +2054,13 @@ func (cs *State) addVote(vote *types.Vote, peerID types.NodeID) (added bool, err
 		return
 	}
 
+	// Verify VoteExtension if precommit
+	if vote.Type == tmproto.PrecommitType {
+		if err = cs.blockExec.VerifyVoteExtension(vote); err != nil {
+			return
+		}
+	}
```

Comment:
Is there any DDoS vector with a validator spamming the same vote multiple times? AFAICS we don't check for duplicates till later and I'm wondering how expensive the verification call might be too make. I would much prefer that we check that this is a unique vote from that validator before doing this verification

Reply:
In that case we have to factor the signature verification logic out from cs.Votes.AddVote, or we could send the extension verification logic as a callback argument. But I agree it would be better to check extension after the signature verification.

Reply:
Ideally we add this check after signature verification and before we add the vote to the VoteSet. I'm not sure what's the best way to do that.

Reply:
For now I think it would be okay to pass an argument `optverify func(Vote) error` to `AddVote` and call it inside

Reply:
Would it be alright to move this into a new issue, and not block this PR on it? (Its mainly a required efficiency improvement)

Just to help pipeline integration testing with it?

Reply:
Sure I don't mind that as long as we don't forget